library/ecp.c

```diff
@@ -3125,10 +3682,76 @@ int mbedtls_ecp_gen_privkey( const mbedtls_ecp_group *grp,
     if( mbedtls_ecp_get_type( grp ) == MBEDTLS_ECP_TYPE_SHORT_WEIERSTRASS )
         return( mbedtls_ecp_gen_privkey_sw( &grp->N, d, f_rng, p_rng ) );
 #endif /* MBEDTLS_ECP_SHORT_WEIERSTRASS_ENABLED */
+#if defined(MBEDTLS_ECP_EDWARDS_ENABLED)
+    if( mbedtls_ecp_get_type( grp ) == MBEDTLS_ECP_TYPE_EDWARDS )
+        return( mbedtls_ecp_gen_privkey_mx( grp->nbits, d, f_rng, p_rng ) );
```

Comment:
This is wrong. The private key in the case of ed25519 is simply "32 octets (256 bits, corresponding to b) of cryptographically secure random data". ed448 only differs by the number of bytes.

Reply:
Actually the RFC only specifies to "simply" generate 32 octets from a true RNG source, but is really wrong using the same approach from Montgomery curves?

Reply:
With x25519 and x448 (which is all we do with Montgomery curves), the private key directly represents an integer in the range [0,p-1] which must be a multiple of 4 or 8, so it's an n-bit number with certain bits masked. With Ed25519 and Ed448, the private key is a random string *k* of a certain length, and when the calculation needs an integer, it uses the bits of hash(*k*) to construct one, not the bits of *k* directly. So the generation mechanism for Edwards and Montgomery curves (or more precisely, for xNNN vs EdNNN, but that's synonymous for us) are completely different.

Reply:
So, would `mbedtls_mpi_fill_random( d, 32, f_rng, p_rng )` will be sufficient then?

Reply:
> So, would `mbedtls_mpi_fill_random( d, 32, f_rng, p_rng )` will be sufficient then?

Yes. The similar bit twiddling work is done after computing the hash in `mbedtls_ecp_expand_ed25519()`, so just random data is enough. You just might want to hardcode it to 32, but compute it from pbits so that it works for ed448 two.

Reply:
> Actually the RFC only specifies to "simply" generate 32 octets from a true RNG source, but is really wrong using the same approach from Montgomery curves?

In addition to the answer from @gilles-peskine-arm, grp->nbits is 0 on Edwards curves, so this function always fail to generate a key.